pkg/operator/controller/controller.go

```diff
@@ -0,0 +1,864 @@
+		rs := createCRDResources()
+		resources = append(resources, rs...)
+	}
+
+	nsrs := createControllerResources(r.getOperatorArgs(cr))
+	resources = append(resources, nsrs...)
+
+	return resources, nil
+}
+
+func createControllerResources(args *OperatorArgs) []runtime.Object {
+	return []runtime.Object{
+		resources.CreateServiceAccount(args.Namespace),
+		resources.CreateControllerRole(args.Namespace),
+		resources.CreateControllerRoleBinding(args.Namespace),
+		resources.CreateControllerDeployment("vm-import-controller", args.Namespace, args.ControllerImage, args.PullPolicy, int32(1)),
+	}
+}
+
+func createCRDResources() []runtime.Object {
+	return []runtime.Object{
+		resources.CreateResourceMapping(),
+		resources.CreateVMImport(),
+	}
+}
+
+func deployClusterResources() bool {
+	return strings.ToLower(os.Getenv("DEPLOY_CLUSTER_RESOURCES")) != "false"
+}
+
+func (r *ReconcileVMImportConfig) watchResourceTypes(resources []runtime.Object) error {
+	types := map[reflect.Type]bool{}
+
+	for _, resource := range resources {
+		t := reflect.TypeOf(resource)
+		if types[t] {
+			continue
+		}
+
+		eventHandler := &handler.EnqueueRequestForOwner{
+			IsController: true,
+			OwnerType:    &vmimportv1alpha1.VMImportConfig{},
+		}
+
+		if err := r.controller.Watch(&source.Kind{Type: resource}, eventHandler); err != nil {
+			if meta.IsNoMatchError(err) {
+				log.Info("No match for type, NOT WATCHING", "type", t)
+				continue
+			}
+			return err
+		}
+
+		log.Info("Watching", "type", t)
+
+		types[t] = true
+	}
+
+	return nil
+}
+
+func setLabel(key, value string, obj metav1.Object) {
+	if obj.GetLabels() == nil {
+		obj.SetLabels(make(map[string]string))
+	}
+	obj.GetLabels()[key] = value
+}
+
+func setLastAppliedConfiguration(obj metav1.Object) error {
+	bytes, err := json.Marshal(obj)
+	if err != nil {
+		return err
+	}
+
+	if obj.GetAnnotations() == nil {
+		obj.SetAnnotations(make(map[string]string))
+	}
+
+	obj.GetAnnotations()[lastAppliedConfigAnnotation] = string(bytes)
+
+	return nil
+}
+
+func sameResource(obj1, obj2 runtime.Object) bool {
+	metaObj1 := obj1.(metav1.Object)
+	metaObj2 := obj2.(metav1.Object)
+
+	if reflect.TypeOf(obj1) != reflect.TypeOf(obj2) ||
+		metaObj1.GetNamespace() != metaObj2.GetNamespace() ||
+		metaObj1.GetName() != metaObj2.GetName() {
+		return false
+	}
+
+	return true
+}
+
+// this is used for testing.  wish this a helper function in test file instead of member
+func (r *ReconcileVMImportConfig) crSetVersion(cr *vmimportv1alpha1.VMImportConfig, version string) error {
+	phase := vmimportv1alpha1.PhaseDeployed
+	if version == "" {
+		phase = vmimportv1alpha1.VMImportPhase("")
+	}
+	cr.Status.ObservedVersion = version
+	cr.Status.OperatorVersion = version
+	cr.Status.TargetVersion = version
+	return r.crUpdate(phase, cr)
+}
+
+func (r *ReconcileVMImportConfig) isMutable(obj runtime.Object) bool {
+	switch obj.(type) {
+	case *corev1.ConfigMap, *corev1.Secret, *rbacv1.RoleBinding, *rbacv1.Role:
```

Comment:
no need to include here ClusterRole and ClusterRoleBinding ?

Reply:
I added them here based on my findings. Without that mutation code fails.